core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java

```diff
@@ -100,54 +103,82 @@ public ManagementContext getManagementContext() {
             return managementContext;
         }
 
+        /**
+         * Scans the bundle being added for a catalog.bom file and adds any entries in it to the catalog.
+         *
+         * @param bundle The bundle being added to the bundle context.
+         * @param bundleEvent The event of the addition.
+         *
+         * @return The items added to the catalog; these will be tracked by the {@link BundleTracker} mechanism
+         *         and supplied to the {@link #removedBundle(Bundle, BundleEvent, Iterable)} method.
+         */
         @Override
-        public Long addingBundle(Bundle bundle, BundleEvent bundleEvent) {
+        public Iterable<? extends CatalogItem<?, ?>> addingBundle(Bundle bundle, BundleEvent bundleEvent) {
+            return scanForCatalog(bundle);
+        }
 
-            final BundleContext bundleContext = FrameworkUtil.getBundle(CatalogBomScanner.class).getBundleContext();
-            if (bundleContext == null) {
-                LOG.info("Bundle context not yet established for bundle {} {} {}", bundleIds(bundle));
-                return null;
-            }
 
-            scanForCatalog(bundle);
+        @Override
+        public void removedBundle(Bundle bundle, BundleEvent bundleEvent, Iterable<? extends CatalogItem<?, ?>> items) {
+            if (!items.iterator().hasNext()) {
+                return;
+            }
+            LOG.debug("Unloading catalog BOM entries from {} {} {}", bundleIds(bundle));
+            List<Exception> exceptions = MutableList.of();
+            final BrooklynCatalog catalog = getManagementContext().getCatalog();
+            for (CatalogItem<?, ?> item : items) {
+                LOG.debug("Unloading {} {} from catalog", item.getSymbolicName(), item.getVersion());
+
+                try {
+                    catalog.deleteCatalogItem(item.getSymbolicName(), item.getVersion());
+                } catch (Exception e) {
+                    LOG.warn("Caught {} unloading {} {} from catalog", new String [] {
+                        e.getMessage(), item.getSymbolicName(), item.getVersion()
+                    });
+                    exceptions.add(e);
+                }
+            }
 
-            return bundle.getBundleId();
+            if (0 < exceptions.size()) {
+                throw new CompoundRuntimeException(
```

Comment:
Hm, I guess I have been assuming that this will just result in a message being logged by Karaf as opposed to it causing a failure - I will have to test that out.